### ISSUE TITLE
Making it easier for people to use...

### ISSUE THREAD

**Issue by mattsah** [defuse/php-encryption#41]
@defuse

I came across your library in some of those slim threads and thought it might be great to use, as I've currently been investigating encryption libraries to finish my security module.

That said, two things I noticed immediately:

1) It lacks composer support
2) It takes over the exception handler (I understand this is for security)

I'd like to fix these two problems and make use of your library, but based on the previous composer.json pull request that is open, I'd like to verify that you're interested first.

With respect to the exception handler, it would make more sense, if the exception handler was changed temporarily during the processes which could throw exceptions.  Your exception handler could be split to a separate class, and http://php.net/manual/en/function.restore-exception-handler.php can be used to restore the previous one once decrypting is done (for example).

**Comment by sarciszewski** [comment]
This is sort of a drop-in library. I could look into forking this to make it composable.

**Comment by defuse** [comment]
I need to do some research to find out how composer works. 

I like the idea of hooking the exception handler only temporarily. I think it's better because it's less likely that the client will have overridden us and replaced our handler with a(nother) vulnerable one.

**Comment by mattsah** [comment]
@sarciszewski 

There's a pull request to make it work with composer.

**Comment by mattsah** [comment]
@defuse 

Please see https://github.com/defuse/php-encryption/pull/42

This removes the exception handler as a separate class and sets it as a simple callback function.  I added two methods to set and reset the exception handler and wrapped the Encrypt and Decrypt methods with them.

It was not clear to me wether or not the CreateNewRandomKey really needed it, I'll leave that up to your judgement.

I also did a bunch of docblock cleanup and generated some simple markdown docs for the API using sage (https://github.com/dotink/sage).

This combined with the composer pull request (which will need some modification since I moved Crypto to an src sub directory for doc purposes) should make this library fairly simple to use for other developers (whether that be the guys from Slim, or myself).

I ran the test script with passing output, and also did a quick modify in the example to test the exception handler was working as implemented (although i did not commit that change as it was an intentional fail).

**Comment by pulyaevskiy** [comment]
@defuse Please see #31 which adds support for composer. I can update it to fit #42 (moving Crypto.php to `src/`).

**Comment by defuse** [comment]
Hang on, I just realized the set-then-restore exception handler pattern won't work. If a crypto exception is correctly caught, the method call to restore the exception handler won't get run, and we'll have permanently changed it. :-(

**Comment by mattsah** [comment]
@defuse Still better than right now, no?  Which is that the exception handler is taken over completely.

**Comment by mattsah** [comment]
That is to say, I agree we should figure out a way around that, but:

1) the register / reset API is internal, so we're not exposing new public API with this
2) we can figure out a means to work around that better... perhaps reverting back to a separate class and on __destruct() it resets the old one?

**Comment by mattsah** [comment]
Not sure how well that will work given __destruct() often isn't called very quickly anyhow.  PHP's GC is relatively slow and often times just comes all at once towards the end of scripts.

PHP 5.5 supports finally, it could be noted that *when* user's handle exceptions they should "finally" restore the the exception handler.  Looking ito if there's some way otherwise to determine if the exception was caught.

**Comment by pulyaevskiy** [comment]
@defuse @mattsah @sarciszewski 

Just wondering if there is a way to avoid handling of uncaught exceptions in the library? It really seems like a job for application that uses this library. 

Though I totally understand security concern about exception stack trace.

There is another way to verify that stack trace is not leaked - just by checking `display_errors` ini setting (should be disabled or 'stderr'). Have you considered this option?

**Comment by mattsah** [comment]
@pulyaevskiy 

Do you recommend completely disabling the library if display errors is enabled?

I'm wondering what would happen if the exceptions themselves try to restore the exception handler... is the exception handler code chosen *before* or after they're instantiated?

**Comment by mattsah** [comment]
Whoops, did not mean to close.

**Comment by pulyaevskiy** [comment]
@mattsah 
> Do you recommend completely disabling the library if display errors is enabled?

Not really. I would rather give an option to enable or disable this check. That would probably require to make this class non-static and pass some sort of `$displayErrorsCheck` flag in the constructor. So for dev environment it would still be possible to see all the stack traces (by passing `$displayErrorsCheck = false`).

**Comment by mattsah** [comment]
@pulyaevskiy @defuse 

It appears as though the __construct() method on the exceptions themselves can restore the old exception handler after the parent::__construct()...

So basically this can happen, each function registers the exception handler, if an exception is thrown, it restores the previous handler in it's __construct() after the parent::__construct(), which means it gets handled by the crypto one, but the old one is restored.

If no exception is thrown, it gets to the end and restores the old one.

**Comment by mattsah** [comment]
@pulyaevskiy 

See: https://github.com/defuse/php-encryption/pull/43

I think this takes care of it.  While I agree that exception handling *should* be something that the app is doing, the point of this is if the person writing the app isn't doing it *properly*, we prevent it from becoming a bigger issue.

As for disabling it.  I wouldn't bother giving people the option if we're sticking with this one.  My only concern would be that there's no guarantee that the except handling recall vs. restoration order wouldn't change.

I suspect it's part of the built in exception's constructor that triggers the setup for the current handler, hence, so long as the parent __construct() is called first, I think we're good.

**Comment by mattsah** [comment]
@pulyaevskiy 

I take it back.  The solution doesn't seem to stick.  I'm not sure what the difference is on different systems, but it's unreliable overall and presumably fragile.

@defuse 

After 2 - 3 hours of trying different things, I'm not sure there's any clean solution.  What I thought was a working solution wasn't.

The long and short of it is, I just don't think this is possible with PHP right now.  It'd be OK, IMO if for PHP 5.5 you registered this and specified that all try/catch's should have a finally { } which resets the exception handler, but for php 5.4 that's not viable either.

Long story short, it could introduce security problems, but if you don't want to own other people's exception handlers, then something else will have to be done.  That said, I can't use this library while it's still taking over exception handling.

**Comment by pulyaevskiy** [comment]
> if the person writing the app isn't doing it properly

I guess I would still be in favor of providing proper API with options for users rather than forcing everyone to tweak the library for their use cases.

First of all, I haven't seen anyone who'd leave `display_errors` enabled on production. It just seems extremely unlikely. Second of all, if someone *do* leave it enabled then he/she already in trouble in terms of security (regardless of how this library works).

So I really believe this library shouldn't care about exception handling.

But, I think it is still very useful to have this `display_errors` check somewhere in `Crypto` + make it configurable and defaulted to `true` so that user is forced to make conscious action to disable it.

**Comment by mattsah** [comment]
@pulyaevskiy 

Well, it's not my library.  It seems pretty clear from the README that security is first concern, which often times forsakes usability.  I gave it my best shot, and couldn't come up with a solution that preserved both... so I'm abandoning this for now.

I'd maintain a separate fork if I thought I was in a position to maintain it from a security standpoint, but I'm not.

If my other solution had worked reliably, I wouldn't think the option is necessary.  I suspect now it's probably the only way, but I don't think that would be considered acceptable in terms of the philosophy.

**Comment by defuse** [comment]
@mattsah I'm curious why the exception handler is stopping you from using it. Are you worried about it making other exceptions in your scripts less reliable? Or is it just the feeling (which I share) that a crypto library shouldn't be taking over something so global?

**Comment by defuse** [comment]
If there was any chance taking over the exception handler could cause other parts of the application to fail, I might prefer to check `display_errors` and refuse to run.

We should quantify the exact risk of not doing the exception handler thing. Which exceptions thrown by this library would result in leaking sensitive information (plaintext or the key).

Maybe there's something else we could do, like recursing down to a depth of 20 function calls before actually raising the exception, so nothing is visible in the stack trace?

**Comment by defuse** [comment]
Does anyone see a reason something like this would be worse or less effective than capturing all exceptions?
    
    function safeThrow($eclass, $msg, $depth = 50)
    {
        if ($depth <=  0) {
            throw new $eclass($msg);
        }
        safeThrow($eclass, $msg, $depth - 1);
    }

Two potential (non-)problems come to mind:

1. It may exhaust PHP's stack. I seriously doubt a depth of `50` would do that, but even if it happens PHP dies with a fatal error about memory exhaustion, leaking no information.

2. It might not be good enough. In the future, PHP might try to 'compress' recursive calls to make the backtrace easier to understand. I doubt that will happen, but in any case we should unit test it. Alternatively, the user might have their stack trace depth set to more than `50` (is that possible in php.ini?) in which case the key would still get leaked. I think that's going to be exceptionally rare, so I don't really care about it, but maybe we should set the depth to something like `1000` just in case.

**Comment by ircmaxell** [comment]
Strong -1 on that. It's hacky and doesn't really solve the problem (it's a bandaid).

The problem here is that an uncaught exception can contain potentially sensitive information in the stacktrace. 

So why not set the exception handler when throwing, and then clear it via the destructor of the exception?

    class CryptoException extends Exception {
        public function __construct($message) {
            parent::__construct($message);
            set_exception_handler(function($e) {
                die("Uncaught crypto exception");
            });
        public function __destruct() {
            restore_exception_handler();
        }
    }

That way it's self contained.

**Comment by pulyaevskiy** [comment]
@ircmaxell @defuse 

I'm not sure I'm entirely following this logic, so I wrote simple test.
Below is CryptoException from @ircmaxell example and a "test" Application class which sets it's own handler.  In my App I have my own exception handler.

```php
<?php // test.php

class CryptoException extends Exception {
    public function __construct($message) {
        parent::__construct($message);
        set_exception_handler(function($e) {
            die("Uncaught crypto exception");
        });
    }
        public function __destruct() {
            restore_exception_handler();
        }
    }



class App
{
    public function __construct()
    {
        set_exception_handler(function($e) {
            die("Caught it!");
        });
    }

    public function run()
    {
        throw new CryptoException('error');
    }
}


$app = new App();

$app->run();
```

Now if I run it I get this:

```bash
% php test.php
Uncaught crypto exception%
```

How can I make my app handling all the exceptions in this case?

AFAIK `throw new Exception()` works something like this:

1. `new Exception()` creates new instance and calls `__construct()` which sets new exception handler
2. `throw` actually throws it. 
3. New exception handlers gets called instead of the one defined in my app.

Please correct me if I'm wrong.

**Comment by defuse** [comment]
Are there any guarantees about when the destructor of the exception will get called? If it is at the whim of the garbage collector, then I see problems.

That solution feels like even more of a hack, since it assumes the user is not keeping the exception object around for very long, and no other exceptions get raised until it gets destructed. It's also harder to test well.

Idea: Can we create an exception object then unset the stack trace before throwing it? Or can we extend an exception class and override the methods for retrieving and printing the stack trace?

**Comment by mattsah** [comment]
It doesn't work in destructor, the garbage collector runs can vary depending on settings, and in smaller scripts it may never get run.

@defuse the reason an exception handler is preventing me from running it is because I have another exception handler set for actual 500 errors.  I don't use libraries where I have to modify the source for each installation.  If I find it indespensible, then I'll maintain a separate fork that doesn't do the exception handling at all, but with composer I expect to be able to require things and be done, not have to go in and start hacking upstream sourcecode.

**Comment by mattsah** [comment]
If it's acceptable to @defuse, I'd vote simply to remove it and let others take the risk of not handling exceptions properly.  But I do understand the security concerns for users who aren't vigilant.

**Comment by defuse** [comment]
The methods of Exception are final, so we can't override them. However, maybe we can take advantage of exception nesting. If we created an exception in a safe context, then nest that one in the exception our library throws, which stack trace will it use? (Typing this on my phone so I can't actually try it.)

**Comment by ircmaxell** [comment]
Let's keep in mind that exception handlers happen at the outer most scope. There's no way to resume from it. It's not like an error handler that can resume. Once the exception handler finishes, it's over for the application.

With that in mind, perhaps we could even tweak it slightly more:

    class CryptoException extends Exception {
        public function __construct($message) {
            parent::__construct($message);
            $original = set_exception_handler(function($e) use (&$original) {
                if ($original) {
                    $original($e);
                }
                error_log("Uncaught crypto exception: " . $e->getMessage());
            });
        }
        public function __destruct() {
            restore_exception_handler();
        }
    }

As far as "the garbage collector runs can vary depending on settings" with respect to destructors, that's not true. The garbage collector (meaning the cyclic collector) is programmable, yes. However that only works with cyclic dependencies (where A has reference to B, and B has reference to A). In the case of normal objects destructors are run via normal reference counting semantics. Which means destructors are always run when the object goes out of scope. So:

     function foo() {
        try {
            throwsException();
        } catch (Exception $e) {
            echo $e->getMessage();
        }
    }

The destructor of `$e` will be called when foo()'s execution ends. 100% of the time. https://github.com/php/php-langspec/blob/bdf57758ce3db17c84b0d3a6fa6d4b877858562b/spec/14-classes.md#destructors

Now, this does fall down if the user "collects" exceptions (meaning they never go out of scope). But with this modification (calling the original), that shouldn't matter as the original is still called in the chain. The problem though is that you shouldn't be collecting exceptions in the first place (that's such a bad practice, but tangential to this discussion).

**Comment by defuse** [comment]
@mattsah Good reason.

**Comment by mattsah** [comment]
@ircmaxell I tested with the __destruct() method the other day and could not get it to work as you describe.  But even if we can ensure this is happening, there's still potential that some code after the encryption processes and within the same scope may throw an additional exception which a user would expect to hit their exception handler.

I'm just, overall, not sure it's a great idea to be taking over (potentially) user set exception handlers.  Also, for dev purposes a stack trace may be useful with respect to the crypto stuff itself.

**Comment by defuse** [comment]
Summing up, there are four paths we could take.

1. Don't try to solve the problem at all.
2. Override the exception handler.
3. Override the exception handler per-exception.
4. The recursion hack.

1 < 2 < 3 < 4.

1 < 2 because then giving the library a corrupt ciphertext could cause it to leak the key to the web or a log file. I won't accept that possibility.

2 < 3 because there is at least one real case we can point to where 2 stopped this library from being used, which probably forces the use of a less secure library, and 3 (if it worked) would make this library work with applications that need to install a handler of their own.

3 < 4 because 3, ignoring the fact that it doesn't work (which I haven't confirmed myself), requires assumptions about how the library is used (not keeping exception objects hanging around), and assumptions about undocumented behavior of the garbage collector. Having to rely on the garbage collector makes 3 extremely hard to test, since the garbage collector may behave differently under different loads (i.e. a real app will be very different from any small test app we write). While 4 is an ugly hack, it makes use of a property that's documented, well-known, and easily testable.

Later I'm going to fool around with instantiating a singleton exception of each type, and trying nested exceptions (I'm not sure if it's the `throw` or the `new` that captures the back trace), and if those fail I'm going with 4, the recursion hack.

**Comment by ircmaxell** [comment]
Well, #3 doesn't rely on the GC. Even if you hold objects around, the parent exception handler will be called. No matter what. The difference is the number of handlers it needs to pop through to get there...

**Comment by mattsah** [comment]
@defuse 

For 1 < 2, I think my problem with this is that the *could* it totally dependent on the framework author.  An author *COULD* also just echo their key out to the world.  An author *COULD* post their key to github in their source code.

As I stated before, the reason I can't accept a globally set and never restored exception handler is because I have my own that serves a purpose, so if your code blew up on my sites, someone would see a nice 500 page in production instead of a stack trace.

This just makes it harder for someone like me to use your library, because you're more concerned with people who don't/won't use it correctly.

And I would argue that those who don't/won't use it correctly, are probably going to be using frameworks or other things which make use of the exception handler anyhow, or at least potentially do, so when your script starts causing some weird behavior and they spend 8 hours trying to figure out where it's coming from, it's this library they'll blame.

The added security of it is nice, but as I said, security and usability often butt heads, and at some point you need to make a decision.  I'm just wondering if such a decision to forsake basically any modern usability for security provided in cases of bad implementation only is a useful trade off.

**Comment by defuse** [comment]
@ircmaxell Oh, I didn't notice you did that! Clever. I'll have to test it myself to see if it works, because it looks very different to what was tried (and didn't work) in #43.

**Comment by defuse** [comment]
Initializing a singleton instance of the exception classes in a safe environment works! I've added a branch with the changes:

https://github.com/defuse/php-encryption/compare/experimental_safeexceptions

In my opinion this is *way* better than 1, 2, 3, or 4. Does anyone see a problem with it?

I only tested it manually, but the backtrace shown has `InitializeSafeExceptions()` as the most recent call, and the `require_once` as the call before it. Exactly as desired. Before I merge this in I'm going to write tests, of course.

**Comment by mattsah** [comment]
@defuse to be clear, this is essentially preventing stack trace without exception handler, yes?  And this being based on that the actual time of calling their intitialization, it was immediatley after the require?  This all still just feels a bit weird, but if it gets around it then so be it.

**Comment by defuse** [comment]
@mattsah Exactly. The exception objects are created immediately after the require, and that's what gets echoed in the stack trace. So the exceptions thrown by this library will have an incorrect (but safe) stack trace, and there's no longer a need to add an exception handler.

**Comment by sarciszewski** [comment]
Hah, that is one clever technique! I don't see any problems. Just need to test in various environments.

**Comment by defuse** [comment]
I added some basic testing: https://github.com/defuse/php-encryption/commit/ac64e80bc0d65ff9405864d2e9c20c905bbd2a36

We've lost the ability to have customized message strings, which are pretty useful for debugging. We can work around that by initializing more exception objects with the strings we want built-in.

**Comment by defuse** [comment]
Hm, tests are failing on Travis. I'll have to find out why later.

**Comment by ircmaxell** [comment]
Strong -1.

Exceptions are not cheap (they involve copying a backtrace, etc). They should not be instantiated until necessary.

Additionally, the lack of a message or a trace in debugging mode (`var_dump($e)` inside of a try/catch) is problematic.

**Comment by defuse** [comment]
Woah, I just tested @ircmaxell's CryptoException that does the recursive popping and it makes PHP segfault. Will investigate more later.

Here is the branch: https://github.com/defuse/php-encryption/tree/per-exception-handler

**Comment by defuse** [comment]
Here is the backtrace

```
Program received signal SIGSEGV, Segmentation fault.
0x000000000066a141 in gc_zval_possible_root ()
(gdb) bt
#0  0x000000000066a141 in gc_zval_possible_root ()
#1  0x00000000006704c7 in zend_objects_destroy_object ()
#2  0x0000000000676c0f in zend_objects_store_del_ref_by_handle_ex ()
#3  0x0000000000676c53 in zend_objects_store_del_ref ()
#4  0x0000000000638480 in _zval_ptr_dtor ()
#5  0x000000000066a515 in ?? ()
#6  0x0000000000676c2c in zend_objects_store_del_ref_by_handle_ex ()
#7  0x0000000000676c53 in zend_objects_store_del_ref ()
#8  0x0000000000638480 in _zval_ptr_dtor ()
#9  0x000000000065b9c6 in ?? ()
#10 0x00000000006f63fe in ?? ()
#11 0x0000000000686100 in execute_ex ()
#12 0x0000000000639fe6 in zend_call_function ()
#13 0x00000000006616f7 in zend_call_method ()
#14 0x00000000006705fb in zend_objects_destroy_object ()
#15 0x0000000000676685 in zend_objects_store_call_destructors ()
#16 0x00000000006387fd in ?? ()
#17 0x0000000000649197 in zend_call_destructors ()
#18 0x00000000005e5ae5 in php_request_shutdown ()
#19 0x00000000006f721d in ?? ()
#20 0x0000000000425560 in ?? ()
#21 0x00007ffff6489800 in __libc_start_main () from /usr/lib/libc.so.6
#22 0x00000000004256a9 in _start ()
```

**Comment by ircmaxell** [comment]
What version of PHP?

**Comment by defuse** [comment]
```
PHP 5.6.6 (cli) (built: Feb 20 2015 17:23:08) 
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologie
```

**Comment by defuse** [comment]
Also segfaults on

```
PHP 5.4.36-0+deb7u3 (cli) (built: Jan  9 2015 08:07:06) 
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies
```

(Debian Wheezy)

**Comment by ircmaxell** [comment]
It's a bug in PHP (obviously). Though not sure how to fix it...

**Comment by sarciszewski** [comment]
Hmm. What is PHP doing under the hood to cause the segfault? (I'm not great at lower-level debugging.) Is this worth opening a report on bugs.php.net or what?

**Comment by defuse** [comment]
Here's a minimal script to reproduce the segfault:

```
<?php

class CryptoException extends Exception {
    public function __construct($message = "") {
        parent::__construct($message);
        $original = set_exception_handler(function($e) use (&$original) {
            if ($original) {
                $original($e);
            }
            error_log("Uncaught crypto exception: " . $e->getMessage());
        });
    }
    public function __destruct() {
        restore_exception_handler();
    }
}

try {
    throw new CryptoException("Hello");
} catch (CryptoException $ex) {
    echo "Caught it.\n";
}
throw new CryptoException("Hello");
```

(For whatever reason both the try-catch and the throw at the end are necessary, if you remove either one it won't segfault).

**Comment by ircmaxell** [comment]
It's a bug in the garbage collector

**Comment by ircmaxell** [comment]
For the record, it works fine in 7 :-)

**Comment by sarciszewski** [comment]
```sh
censored:~ # php blah.php
Caught it.
Uncaught crypto exception: Hello
Uncaught crypto exception: Hello
censored:~ # php -v
PHP 5.3.17 (cli)
Copyright (c) 1997-2012 The PHP Group
Zend Engine v2.3.0, Copyright (c) 1998-2012 Zend Technologies
```

**Comment by defuse** [comment]
Okay, so given that that won't work, we're left with either the recursion hack or pre-instantiated exceptions.

Arguments for the recursion hack: The stack trace isn't outright lying, the entire trace is still available via `var_dump($ex)` (including keys), and it's not wasting resources until an exception needs to be thrown (but then it's generating a huge exception).

Arguments for pre-instantiated exceptions: The exception thrown won't contain the key at all. I don't know PHP internals all that well, but when they get instantiated they're only a few methods deep (`require_once` then two of our methods), so they won't be huge, and we only need ~5 of them. But maybe they're really slow to create...

**Comment by sarciszewski** [comment]
And then we have the nuclear option: Just use `die()` instead of exceptions. The library won't operate in badly configured environments anyway.

**Comment by defuse** [comment]
There's legitimate uses of exceptions, for example to inform the client that the ciphertext was tampered with, so `die()`ing instead of raising those could cause vulnerabilities (or could cause attacks to go un-noticed).

**Comment by wheelsandcogs** [comment]
I really think this is a "Don't try to solve the problem at all" issue.

If the stack trace displays the key in a development environment then it's a non-issue because the developer already has access to the key.

If stack traces are being displayed in a production environment then the developer has bigger problems than leaking a key! As long as this is clear in the documentation then I can't see a need for any special exception handling.

**Comment by defuse** [comment]
I agree, I think we should stop trying to override the exception handler. There's no good solution so we should just document the risk.

**Comment by defuse** [comment]
Opened #97 to remove the exception handler override. Now this issue can be closed.
